google-cloud-bigquerystorage/src/main/java/com/google/cloud/bigquery/storage/v1/JsonStreamWriter.java

```diff
@@ -105,6 +105,43 @@ public ApiFuture<AppendRowsResponse> append(JSONArray jsonArr)
     return append(jsonArr, -1);
   }
 
+  private void refreshWriter(TableSchema updatedSchema)
+      throws DescriptorValidationException, IOException {
+    Preconditions.checkNotNull(updatedSchema, "updatedSchema is null.");
+    LOG.info("Refresh internal writer due to schema update");
+    // Close the StreamWriter
+    this.streamWriter.close();
+    // Update JsonStreamWriter's TableSchema and Descriptor
+    this.tableSchema = updatedSchema;
+    this.descriptor =
+        BQTableSchemaToProtoDescriptor.convertBQTableSchemaToProtoDescriptor(updatedSchema);
+    this.protoSchema = ProtoSchemaConverter.convert(this.descriptor);
+    this.totalMessageSize = protoSchema.getSerializedSize();
+    // Create a new underlying StreamWriter with the updated TableSchema and Descriptor
+    this.streamWriter = streamWriterBuilder.setWriterSchema(this.protoSchema).build();
+  }
+
+  private Message buildMessage(JSONObject json)
+      throws InterruptedException, DescriptorValidationException, IOException {
+    try {
+      return JsonToProtoMessage.convertJsonToProtoMessage(
+          this.descriptor, this.tableSchema, json, ignoreUnknownFields);
+    } catch (Exceptions.JsonDataHasUnknownFieldException ex) {
+      // Backend cache for GetWriteStream schema staleness can be 30 seconds, wait a bit before
+      // trying to get the table schema to increase the chance of succeed. This is to avoid
+      // client's invalid data caused storm of GetWriteStream.
+      Thread.sleep(3500);
```

Comment:
I don't have enough understanding of the underlying staleness issue, but my hunch is that explicit Thread.sleep() within the library is going to mask plenty of weird errors. 

It's a bigger change, but consider allowing the user to supply a RetryPolicy as part of the builder, and then execute this convertJsonToProtoMessage() based on the user supplied policy. It gives the users more flexibility to iterate over different behaviors, i.e try more than just once, supply hard timeout, etc.

If this is out of scope in your opinion, then at least I would expect to see a LOG.warning before entering 3.5seconds of sleep.